File: flutter/lib/core/runner.dart
import 'dart:async';
import 'channel.dart';
import 'commands.dart';
import 'types.dart';

/// Serializes structured surge commands over a single terminal channel, framing
/// each with sentinel markers so output and exit codes can be captured
/// deterministically. Mirror of `packages/shared/src/runner.ts`.
class CommandRunner {
  CommandRunner(this._channel, this._profile) {
    _sub = _channel.output.listen(_ingest);
  }

  static const _begin = '__SM_BEGIN__';
  static const _end = '__SM_END__';

  final TerminalChannel _channel;
  final SurgeProfile _profile;
  late final StreamSubscription<String> _sub;

  final List<_QueuedCommand> _queue = [];
  _QueuedCommand? _active;
  String _buffer = '';
  int _seq = 0;
  Timer? _timer;

  Future<CommandResult> run(
    SurgeAction action, [
    List<String> args = const [],
    Duration timeout = const Duration(seconds: 15),
  ]) {
    final commandLine = buildCommandLine(_profile, action, args);
    final id = '${DateTime.now().millisecondsSinceEpoch.toRadixString(36)}_'
        '${(_seq++).toRadixString(36)}';
    final completer = Completer<CommandResult>();
    _queue.add(_QueuedCommand(
      id: id,
      action: action,
      commandLine: commandLine,
      completer: completer,
      timeout: timeout,
    ));
    _pump();
    return completer.future;
  }

  void _pump() {
    if (_active != null || _queue.isEmpty) return;
    final next = _queue.removeAt(0);
    _active = next;
    next.startedAt = DateTime.now();
    _buffer = '';
    final framed = "printf '\\n$_begin %s\\n' '${next.id}'; "
        '${next.commandLine}; '
        "__sm_rc=\$?; printf '\\n$_end %s %s\\n' '${next.id}' \"\$__sm_rc\"\n";
    _channel.write(framed);
    _timer = Timer(next.timeout, _onTimeout);
  }

  void _onTimeout() {
    final cmd = _active;
    if (cmd == null) return;
    _active = null;
    _timer = null;
    cmd.completer.completeError(
      TimeoutException('Command "${cmd.action.name}" timed out'),
    );
    _pump();
  }

  void _ingest(String chunk) {
    final cmd = _active;
    if (cmd == null) return;
    _buffer += chunk;
    if (_buffer.length > 8 * 1024 * 1024) {
      _buffer = _buffer.substring(_buffer.length - 4 * 1024 * 1024);
    }

    final endRe = RegExp('$_end ${cmd.id} (-?\\d+)');
    final endMatch = endRe.firstMatch(_buffer);
    if (endMatch == null) return;

    final beginMarker = '$_begin ${cmd.id}';
    final beginIdx = _buffer.indexOf(beginMarker);
    final captureStart =
        beginIdx == -1 ? 0 : _buffer.indexOf('\n', beginIdx) + 1;
    final captureEnd = _buffer.lastIndexOf('\n', endMatch.start);
    final stdout = _stripEcho(
      _buffer.substring(captureStart, captureEnd < 0 ? endMatch.start : captureEnd),
      cmd.commandLine,
    );
    final exitCode = int.tryParse(endMatch.group(1)!) ?? -1;

    _timer?.cancel();
    _timer = null;
    _active = null;
    cmd.completer.complete(CommandResult(
      action: cmd.action,
      exitCode: exitCode,
      stdout: stdout,
      durationMs: DateTime.now().difference(cmd.startedAt!).inMilliseconds,
    ));
    _pump();
  }

  String _stripEcho(String captured, String commandLine) {
    var out = captured;
    final echoIdx = out.indexOf(commandLine);
    if (echoIdx != -1) {
      final after = out.indexOf('\n', echoIdx);
      if (after != -1) out = out.substring(after + 1);
    }
    return out.replaceFirst(RegExp(r'^\r?\n'), '').replaceFirst(RegExp(r'\r?\n\s*$'), '');
  }

  Future<void> dispose() async {
    await _sub.cancel();
    _timer?.cancel();
    final err = StateError('Runner disposed');
    _active?.completer.completeError(err);
    for (final q in _queue) {
      q.completer.completeError(err);
    }
    _queue.clear();
    _active = null;
  }
}

class _QueuedCommand {
  _QueuedCommand({
    required this.id,
    required this.action,
    required this.commandLine,
    required this.completer,
    required this.timeout,
  });

  final String id;
  final SurgeAction action;
  final String commandLine;
  final Completer<CommandResult> completer;
  final Duration timeout;
  DateTime? startedAt;
}

File: packages/shared/src/runner.ts
import { buildCommandLine } from "./commands.js";
import type { CommandResult, SurgeAction, SurgeProfile } from "./types.js";

/**
 * Minimal duplex view of a terminal session (mosh PTY or SSH shell). The runner
 * only needs to write bytes and subscribe to output; the concrete transport
 * (node-pty, ssh2 channel, Dart mosh client) implements this.
 */
export interface TerminalChannel {
  write(data: string): void;
  /** Subscribe to raw output. Returns an unsubscribe function. */
  onData(listener: (chunk: string) => void): () => void;
}

const BEGIN = "__SM_BEGIN__";
const END = "__SM_END__";

interface QueuedCommand {
  id: string;
  action: SurgeAction;
  commandLine: string;
  resolve: (r: CommandResult) => void;
  reject: (e: Error) => void;
  timeoutMs: number;
}

/**
 * Serializes structured surge commands over a single interactive terminal,
 * framing each with sentinel markers so output and exit codes can be captured
 * deterministically even though the channel is a raw byte stream.
 *
 * Commands are queued and run one at a time so markers never interleave.
 */
export class CommandRunner {
  private buffer = "";
  private queue: QueuedCommand[] = [];
  private active: QueuedCommand | null = null;
  private activeStart = 0;
  private timer: ReturnType<typeof setTimeout> | null = null;
  private readonly unsubscribe: () => void;
  private seq = 0;
  private readonly channel: TerminalChannel;
  private readonly profile: SurgeProfile;

  constructor(channel: TerminalChannel, profile: SurgeProfile) {
    this.channel = channel;
    this.profile = profile;
    this.unsubscribe = channel.onData((chunk) => this.ingest(chunk));
  }

  dispose(): void {
    this.unsubscribe();
    if (this.timer) clearTimeout(this.timer);
    const err = new Error("Runner disposed");
    if (this.active) this.active.reject(err);
    for (const q of this.queue) q.reject(err);
    this.queue = [];
    this.active = null;
  }

  run(
    action: SurgeAction,
    args: string[] = [],
    timeoutMs = 15_000,
  ): Promise<CommandResult> {
    const commandLine = buildCommandLine(this.profile, action, args);
    const id = `${Date.now().toString(36)}_${(this.seq++).toString(36)}`;
    return new Promise<CommandResult>((resolve, reject) => {
      this.queue.push({ id, action, commandLine, resolve, reject, timeoutMs });
      this.pump();
    });
  }

  private pump(): void {
    if (this.active || this.queue.length === 0) return;
    const next = this.queue.shift()!;
    this.active = next;
    this.activeStart = Date.now();
    this.buffer = "";
    // Wrap so we can find begin/end markers and capture the real exit code.
    // The leading newline guards against a partial prompt on the same line.
    const framed =
      `printf '\\n${BEGIN} %s\\n' '${next.id}'; ` +
      `${next.commandLine}; ` +
      `__sm_rc=$?; printf '\\n${END} %s %s\\n' '${next.id}' "$__sm_rc"\n`;
    this.channel.write(framed);
    this.timer = setTimeout(() => this.onTimeout(), next.timeoutMs);
  }

  private onTimeout(): void {
    const cmd = this.active;
    if (!cmd) return;
    this.active = null;
    this.timer = null;
    cmd.reject(new Error(`Command "${cmd.action}" timed out`));
    this.pump();
  }

  private ingest(chunk: string): void {
    if (!this.active) return; // not awaiting a structured command
    this.buffer += chunk;
    // Cap buffer growth from a runaway/streaming command.
    if (this.buffer.length > 8 * 1024 * 1024) {
      this.buffer = this.buffer.slice(-4 * 1024 * 1024);
    }

    const cmd = this.active;
    const beginMarker = `${BEGIN} ${cmd.id}`;
    const endRe = new RegExp(`${END} ${cmd.id} (-?\\d+)`);
    const endMatch = endRe.exec(this.buffer);
    if (!endMatch) return;

    const beginIdx = this.buffer.indexOf(beginMarker);
    const captureStart =
      beginIdx === -1 ? 0 : this.buffer.indexOf("\n", beginIdx) + 1;
    const captureEnd = this.buffer.lastIndexOf("\n", endMatch.index);
    const stdout = stripEcho(
      this.buffer.slice(captureStart, captureEnd < 0 ? endMatch.index : captureEnd),
      cmd.commandLine,
    );
    const exitCode = Number(endMatch[1]);

    if (this.timer) clearTimeout(this.timer);
    this.timer = null;
    this.active = null;
    const result: CommandResult = {
      action: cmd.action,
      exitCode,
      stdout,
      durationMs: Date.now() - this.activeStart,
    };
    cmd.resolve(result);
    this.pump();
  }
}

/**
 * Remove the echoed command line(s) a PTY reflects before the real output, plus
 * trailing shell-prompt noise.
 */
function stripEcho(captured: string, commandLine: string): string {
  let out = captured;
  const echoIdx = out.indexOf(commandLine);
  if (echoIdx !== -1) {
    const after = out.indexOf("\n", echoIdx);
    if (after !== -1) out = out.slice(after + 1);
  }
  return out.replace(/^\r?\n/, "").replace(/\r?\n\s*$/, "");
}
